### PR TITLE
Fixed #27452 -- Added serial fields to django.contrib.postgres

### PR DESCRIPTION
**Pull request by codingjoe**
https://code.djangoproject.com/ticket/27452

### CODE REVIEW

**Comment by charettes** [comment]
I'm not sure I understand the advantage of these fields over `AutoField` and `BigAutoField` which are also using the `serial` and `bigserial` types?

https://github.com/django/django/blob/c1b94e32fb3df25d72b5e9973da7928dddbc3a2e/django/db/backends/postgresql/base.py#L74-L75

In the light of `serial` being somewhat deprecated in favor or identity columns [which the aforementioned fields are planned to be to switched on PostgreSQL](https://code.djangoproject.com/ticket/30511) I guess adding `contrib.postgres.SerialField` and friends could make more sense in this context.

**Comment by codingjoe** [comment]
@charettes I believe this is a question you might want to ask on the ticket on track, rather than here on the unfinished patch. Which happens to be reopened and the original PR was reviewed by you in the past, see also #7525 

To answer your question, yes this was meant as a way to add serial fields as non-primary keys. That being said, especially with rise of `IDENTITY` fields, it's nice to be still able to use serials.
We sure can add non-pk identity fields too, once the database backends actually support the `GENERATED ALWAYS` statement.

**Comment by pope1ni** [comment]
@codingjoe I'm adding support for `SmallAutoField` in #11611 which is backed by `smallserial` on PostgreSQL. It would make sense for this to also add `SmallSerialField`.

**Comment by codingjoe** [comment]
> @codingjoe I'm adding support for `SmallAutoField` in #11611 which is backed by `smallserial` on PostgreSQL. It would make sense for this to also add `SmallSerialField`.

Thanks @pope1ni the patch already includes a `SmallSerialField` :)

**Comment by pope1ni** [comment]
> Thanks @pope1ni the patch already includes a `SmallSerialField` :)

Hah, so it does!

File: django/db/backends/postgresql/features.py [hunk 1]
```diff
@@ -55,6 +55,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_aggregate_filter_clause = True
     supported_explain_formats = {'JSON', 'TEXT', 'XML', 'YAML'}
     validates_explain_options = False  # A query will error on invalid options.
+    has_native_serial_field = True
```
Comment (pope1ni) [inline django/db/backends/postgresql/features.py]: Please position this up near `has_native_uuid_field` and `has_native_duration_field`.

You also need to add `has_native_serial_field = False` to `django/db/backends/base/features.py`.

File: docs/ref/contrib/postgres/fields.txt [hunk 2]
```diff
@@ -19,6 +19,43 @@ may be a good choice for the :ref:`range fields <range-fields>` and
 :class:`HStoreField`, and GIN may be helpful for :class:`ArrayField` and
 :class:`JSONField`.
 
+``SerialField``
+===============
+
+.. versionadded:: 2.2
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: `.. versionadded:: 3.0`

Comment (pope1ni) [reply docs/ref/contrib/postgres/fields.txt]: Fields are documented in alphabetical order. I would also suggest combining these into a single ` ``SerialField``s ` heading, similar to ` ``CIText`` fields `.

File: docs/ref/contrib/postgres/fields.txt [hunk 3]
```diff
@@ -19,6 +19,43 @@ may be a good choice for the :ref:`range fields <range-fields>` and
...
+    .. attribute:: base_field
+
+        This is an optional argument.
+
+        The field is unique by default, however a serial does not need to be unique.
+
+``BigSerialField``
+==================
+
+.. versionadded:: 2.2
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: `.. versionadded:: 3.0`

File: docs/ref/contrib/postgres/fields.txt [hunk 4]
```diff
@@ -19,6 +19,43 @@ may be a good choice for the :ref:`range fields <range-fields>` and
...
+
+.. class:: BigSerialField(unique=True)
+
+    A 8-byte integer, much like an :class:`SerialField` except that it is
+    guaranteed to fit numbers from ``1`` to ``9223372036854775807``.
+
+``SmallSerialField``
+====================
+
+.. versionadded:: 2.2
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: `.. versionadded:: 3.0`

File: docs/releases/3.0.txt [hunk 5]
```diff
@@ -133,7 +133,10 @@ Minor features
 :mod:`django.contrib.postgres`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* The new :class:`~django.contrib.postgres.fields.SerialField` and
+  :class:`~django.contrib.postgres.fields.BigSerialField`. Much like the
```
Comment (pope1ni) [inline docs/releases/3.0.txt]: Don't forget to mention `SmallSerialField` here...

File: tests/postgres_tests/test_serialfield.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        self.assertIn(small.get_internal_type(), PostgresDatabaseWrapper.data_types)
+        self.assertIn(big.get_internal_type(), PostgresDatabaseWrapper.data_types)
+
+        self.assertEqual(PostgresDatabaseWrapper.data_types[regular.get_internal_type()], 'serial')
+        self.assertEqual(PostgresDatabaseWrapper.data_types[small.get_internal_type()], 'smallserial')
+        self.assertEqual(PostgresDatabaseWrapper.data_types[big.get_internal_type()], 'bigserial')
+
+
+@skipUnlessDBFeature('has_native_serial_field')
+class SerialFieldTests(PostgreSQLTestCase):
```
Comment (pope1ni) [inline tests/postgres_tests/test_serialfield.py]: I would suggest subclassing `IntegerFieldTests` to ensure that tests for validators, etc. are run for these fields too.

See changes to `tests/model_fields/test_autofield.py` in outstanding PR #10680 for an example.

This test should also be duplicated for a `BigSerialModel` and `SmallSerialModel`.

Comment (codingjoe) [reply tests/postgres_tests/test_serialfield.py]: I can't import that unit test, since it's from another app and imports models that are not loaded :/

Comment (pope1ni) [reply tests/postgres_tests/test_serialfield.py]: Ah, yes. I guess the tests mentioned will just need to be copied...

File: django/contrib/postgres/expressions.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+class Default(BaseExpression):
+    """
+    Expression for ``DEFAULT``.
+
+    In an insert query this will return the database default.
+    """
+
+    def as_sql(self, compiler, connection):
+        return connection.ops.pk_default_value(), []
```
Comment (pope1ni) [inline django/contrib/postgres/expressions.py]: If we make this available for all backends, we should probably make some changes in this area:

- We should rename `pk_default_value()` to `default_value()`.
- We should have a `pk` argument to that so we can do `return 'NULL' if pk else 'DEFAULT'` for Oracle in that method.

File: docs/ref/contrib/postgres/fields.txt [hunk 8]
```diff
@@ -951,3 +951,40 @@ used with a custom range functions that expected boundaries, for example to
 define :class:`~django.contrib.postgres.constraints.ExclusionConstraint`. See
 `the PostgreSQL documentation for the full details <https://www.postgresql.org/
 docs/current/rangetypes.html#RANGETYPES-INCLUSIVITY>`_.
+
+``Serial`` fields
+=================
+
+.. versionadded:: 3.1
+
+.. class:: SerialField(unique=True, editable=False)
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: This should have `, **options` at the end of the arguments.

File: docs/ref/contrib/postgres/fields.txt [hunk 9]
```diff
@@ -951,3 +951,40 @@ used with a custom range functions that expected boundaries, for example to
...
+
+    .. attribute:: editable
+
+        This is an optional argument.
+
+        The field is not edtiable by default, similar to the
+        :class:`AutoField <django.db.models.AutoField>`, since its value is set
+        by the database. However, it can be changed back, to be editable.
+
+.. class:: BigSerialField(unique=True)
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: This should have `, editable=False, **options` at the end of the arguments.

File: docs/ref/contrib/postgres/fields.txt [hunk 10]
```diff
@@ -951,3 +951,40 @@ used with a custom range functions that expected boundaries, for example to
...
+
+    A 8-byte integer, much like an :class:`SerialField` except that it is
+    guaranteed to fit numbers from ``1`` to ``9223372036854775807``.
+
+.. class:: SmallSerialField(unique=True)
```
Comment (pope1ni) [inline docs/ref/contrib/postgres/fields.txt]: This should have `, editable=False, **options` at the end of the arguments.

**Comment by pope1ni** [comment]
Can you also update the commit message and PR name to:

`Fixed #27452 -- Added serial fields to django.contrib.postgres.`

Thanks!

**Comment by codingjoe** [comment]
buildbot, test on oracle.

**Comment by codingjoe** [comment]
Point 2 is now address in a separate ticket https://code.djangoproject.com/ticket/30969

I also rebased.

@pope1ni would you be willing to review this one more time 🙏 I'd really like to wrap this up.

**Comment by codingjoe** [comment]
buildbot, test on oracle.

File: tests/postgres_tests/fields.py [hunk 11]
```diff
@@ -4,16 +4,20 @@
...
-        ArrayField, BigIntegerRangeField, CICharField, CIEmailField,
+        ArrayField, BigIntegerRangeField, BigSerialField, CICharField, CIEmailField,
         CITextField, DateRangeField, DateTimeRangeField, DecimalRangeField,
-        HStoreField, IntegerRangeField, JSONField,
+        HStoreField, IntegerRangeField, JSONField, SmallSerialField, SerialField
     )
     from django.contrib.postgres.search import SearchVectorField
-except ImportError:
+except (ImportError, ImproperlyConfigured):
+    from django.db.backends.dummy.base import DatabaseWrapper  # NOQA
```
Comment (felixxm) [inline tests/postgres_tests/fields.py]: This change looks unrelated.

File: django/contrib/postgres/fields/serial.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+class SerialFieldMixin:
+    db_returning = True
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **{
+            'editable': False,
+            **kwargs,
+            'blank': True,
+            'default': Default(),
```
Comment (felixxm) [inline django/contrib/postgres/fields/serial.py]: We should move introducing `Default()` expression to a separate commit.

File: django/db/models/__init__.py [hunk 13]
```diff
@@ -40,7 +41,7 @@
     'ObjectDoesNotExist', 'signals',
     'CASCADE', 'DO_NOTHING', 'PROTECT', 'RESTRICT', 'SET', 'SET_DEFAULT',
     'SET_NULL', 'ProtectedError', 'RestrictedError',
-    'Case', 'Exists', 'Expression', 'ExpressionList', 'ExpressionWrapper', 'F',
+    'Case', 'Default', 'Exists', 'Expression', 'ExpressionList', 'ExpressionWrapper', 'F',
```
Comment (felixxm) [inline django/db/models/__init__.py]: Wrap at 79 chars.

File: docs/ref/models/expressions.txt [hunk 14]
```diff
@@ -705,6 +705,21 @@ should avoid them if possible.
...
+
+.. class:: Default()
+
+The default expression is used on ``INSERT`` or ``UPDATE`` statements and lets
+the database decide what value to insert or set. The default clause usually
+requires a default constraint for a column to be defined in the database
+schema.
+
+The default expression is usually used to set the primary key, if the primary
+key is an :class:`~django.db.models.AutoField`.
```
Comment (felixxm) [inline docs/ref/models/expressions.txt]: IMO, we shouldn't document `Default` because it's a part of internal API and users may have an impression that we support db defaults.

**Comment by InvalidInterrupt** [comment]
What's the rationale for forcing Serial Fields to be unique? Sorry if this has been answered; I couldn't find it in the discussion.

File: django/contrib/postgres/fields/serial.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+            'unique': True,
+        })
```
Comment (InvalidInterrupt) [inline django/contrib/postgres/fields/serial.py:18]: `serial` columns in PostgreSQL have a `NOT NULL` constraint. You may want to force `null` to be `False`

File: docs/ref/contrib/postgres/fields.txt [hunk 16]
```diff
@@ -1007,3 +1007,37 @@ used with a custom range functions that expected boundaries, for example to
...
+=================
+
+.. versionadded:: 3.1
+
+.. class:: SerialField(unique=True, editable=False, **options)
+
+    A field similar to :class:`AutoField <django.db.models.AutoField>`
+    except it does not need to be a primary key. It uses a separate postgres
+    sequence to automatically increment the integer value. It stores a 4-byte
+    integer that fits numbers from ``1`` to ``2147483647``.
```
Comment (InvalidInterrupt) [inline docs/ref/contrib/postgres/fields.txt:1021]: While the backing sequence will never use them, zero and negative numbers can be stored when the field is editable. This last sentence seems to suggest otherwise. I have the same concern with the docs for `BigSerialField` and `SmallSerialField`.

**Comment by codingjoe** [comment]
I will close this for now, due to a lack of interest. Since Django does support returning values since v3.0 everyone who needs such a field can build it. Yet, I don't have the time get the feature into Django for now.

**Comment by InvalidInterrupt** [comment]
I've opened a new PR based on this work: #12989
